bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestHeapSizeManager.java

```diff
@@ -0,0 +1,182 @@
+
+    long id2 = underTest.registerOperationWithHeapSize(1);
+    assertTrue(underTest.hasInflightRequests());
+    assertTrue(underTest.isFull());
+
+    underTest.markOperationCompleted(id);
+    assertFalse(underTest.isFull());
+    assertTrue(underTest.hasInflightRequests());
+
+    underTest.markOperationCompleted(id2);
+    assertFalse(underTest.isFull());
+    assertFalse(underTest.hasInflightRequests());
+}
+
+  @Test
+  public void testSize() throws InterruptedException{
+    HeapSizeManager underTest = new HeapSizeManager(10l, 1000, executorService);
+    long id = underTest.registerOperationWithHeapSize(5l);
+    assertTrue(underTest.hasInflightRequests());
+    assertFalse(underTest.isFull());
+    assertEquals(5l, underTest.getHeapSize());
+
+    long id2 = underTest.registerOperationWithHeapSize(4l);
+    assertTrue(underTest.hasInflightRequests());
+    assertFalse(underTest.isFull());
+    assertEquals(9l, underTest.getHeapSize());
+
+    long id3 = underTest.registerOperationWithHeapSize(1l);
+    assertTrue(underTest.hasInflightRequests());
+    assertTrue(underTest.isFull());
+    assertEquals(10l, underTest.getHeapSize());
+
+    underTest.markOperationCompleted(id);
+    assertFalse(underTest.isFull());
+    assertEquals(5l, underTest.getHeapSize());
+
+    underTest.markOperationCompleted(id2);
+    underTest.markOperationCompleted(id3);
+    assertFalse(underTest.hasInflightRequests());
+  }
+
+  @Test
+  public void testCallback() throws InterruptedException {
+    final List<Runnable> runnables = new ArrayList<>();
+    Mockito.doAnswer(new Answer<Void>() {
+      @Override
+      public Void answer(InvocationOnMock invocation) throws Throwable {
+        runnables.add((Runnable) invocation.getArguments()[0]);
+        return null;
+      }
+    }).when(future).addListener(Matchers.any(Runnable.class), Matchers.eq(executorService));
+
+    HeapSizeManager underTest = new HeapSizeManager(10l, 1000, executorService);
+    long id = underTest.registerOperationWithHeapSize(5l);
+    assertTrue(underTest.hasInflightRequests());
+
+    underTest.addCallback(future, id);
+    assertFalse(runnables.isEmpty());
+    assertTrue(underTest.hasInflightRequests());
+
+    for (Runnable runnable : runnables) {
+      runnable.run();
+    }
+    assertFalse(underTest.hasInflightRequests());
+  }
+
+  @Test
+  public void testSizeLimitReachWaits() throws InterruptedException {
+    ExecutorService pool = Executors.newCachedThreadPool();
+    try {
+      final HeapSizeManager underTest = new HeapSizeManager(1l, 1, executorService);
+      long id = underTest.registerOperationWithHeapSize(5l);
+      assertTrue(underTest.hasInflightRequests());
+      final AtomicBoolean secondRequestRegistered = new AtomicBoolean();
+      pool.submit(new Runnable() {
+        @Override
+        public void run() {
+          try {
+            underTest.registerOperationWithHeapSize(5l);
+            secondRequestRegistered.set(true);
```

Comment:
isFull() is true before the runnable is submitted.  The max heap size value is set to 1l here.  The first registration took all of the quota.